feat!: add multiple custom assets

hal3e opened the pull request:
closes: https://github.com/FuelLabs/fuels-rs/issues/475

Initially, the idea was to have something like this:
`WalletsConfig::new_multiple_assets([(asset_id_1, 23), (asset_id_2, 42), ...])`
However, we have decided that the user should also be able to set the number of coins and the amount per coin. So we have the following new method:
`WalletsConfig::new_multiple_assets([(asset_id_1, 100, 23), (asset_id_2, 100, 42), ...])` where we specify both the number of coins and the amount per coin in the tuple.

I have:
- refactored the `WalletConfig` and implemented the new method
- added the underlying function to create the coins and added a unit test
- added a test with the provider to see if the assets are generated and if we can retrieve them
- added an example and updated the docs

Regarding the approach. There is an open question whether we should create the base asset automatically or leave it to the user. I went with the second option. If the user is defining custom assets then maybe he also wants to be able to specify the number of coins and amount of the base asset. I have added this to the docs. I would like to hear you opinion on this.

hal3e commented:
In the last commit I did some refactoring of the test and removed unnecessary lines and increased the number of wallets. I had it set to 1 for debugging.

vnepveu commented on examples/wallets/src/lib.rs:
```diff
@@ -189,6 +189,43 @@ mod tests {
 
+    #[tokio::test]
+    #[allow(unused_variables)]
+    async fn setup_wallet_custom_assets() -> Result<(), Box<dyn std::error::Error>> {
```
I think you can use `Result<(), Error>` here

vnepveu commented on docs/src/wallets/test-wallets.md:
```diff
@@ -27,6 +27,17 @@ You can create a test wallet containing multiple assets (including the base asse
 - coins: `Vec<(UtxoId, Coin)>` has num_assets * coins_per_assets coins (UTXOs)
 - asset_ids: `Vec<AssetId>` contains the num_assets randomly generated `AssetId`s (always includes the base asset)
 
+## Setting up a test wallet with multiple custom assets
```
I think you should also rename the previous section with `with multiple random assets` or something.

vnepveu commented on examples/wallets/src/lib.rs:
```diff
@@ -189,6 +189,43 @@ mod tests {
+        let coins_per_asset_2 = 10;
+        let amount_per_coin_2 = 12;
+
+        let assets = &[
+            (asset_id_base, coins_per_asset_base, amount_per_coin_base),
+            (asset_id_1, coins_per_asset_1, amount_per_coin_1),
+            (asset_id_2, coins_per_asset_2, amount_per_coin_2),
+        ];
```
I feel that using `Vec` here would be cleaner. Also, maybe a struct would be in order here? Something like `AssetSetup` with the 3 fields?

hal3e replied:
This was definitely on my mind as you need to know the order. However the requested function was a vector with tuples. If we agree on the change I will gladly implement it.

vnepveu replied:
Ah you're right, let's ping @digorithm to see what he thinks of a struct. the struct could be built using helper functions taki

vnepveu commented on packages/fuels-test-helpers/src/lib.rs:
```diff
@@ -91,6 +91,18 @@ pub fn setup_multiple_assets_coins(
+/// Create a vector of of UTXOs with the provided AssetIds, num_coins and amount_per_coin
+pub fn setup_custom_assets_coins(
+    owner: Address,
+    assets: &[(AssetId, u64, u64)],
```
Using `Vec` is preferred afaikk

vnepveu commented on packages/fuels-test-helpers/src/wallets_config.rs:
```diff
@@ -19,16 +20,18 @@ impl WalletsConfig {
     }
 
-    pub fn new_single(coins: Option<u64>, amount: Option<u64>) -> Self {
+    pub fn new_multiple_assets(num_wallets: u64, assets: Vec<(AssetId, u64, u64)>) -> Self {
         Self {
-            num_wallets: 1,
-            coins_per_wallet: coins.unwrap_or(DEFAULT_NUM_COINS),
-            coin_amount: amount.unwrap_or(DEFAULT_COIN_AMOUNT),
```
I don't think this setup function should disappear, it's quite practical to be able to just generate a new single wallet without having to specify assets.

hal3e replied:
It is basically the same as `new` but the first parameter is set to 1 and it was used once...

vnepveu replied:
Ok I thought it was more used than that. I just always want make sure that we keep very simple and easy defaults available.

vnepveu reviewed (changes requested):
It's looking good, I think you can use a struct and `vec!` more, also I'm not sure why you removed the `setup_single` function.

hal3e commented:
@vnepveu thanks for the review! In the end, I have added the `AssetConfig` struct and refactored the `new_multiple_assets` function. I like how it looks now. Let's see what others say :)

MujkicA commented on docs/src/wallets/test-wallets.md:
````diff
@@ -27,6 +27,17 @@ You can create a test wallet containing multiple assets (including the base asse
+{{#include ../../../examples/wallets/src/lib.rs:custom_assets_wallet}}
+```
+
+>**Note** In this case you need to specify also the base asset and the corresponding number of
````
```suggestion
>**Note** In this case, you need to specify also the base asset and the corresponding number of
```

MujkicA reviewed (dismissed):
LGTM :ok_hand:

hal3e commented:
Thanks @MujkicA for the review!

digorithm commented:
Branch outdated -- could you sync with base? 😁

digorithm commented on docs/src/wallets/test-wallets.md:
```diff
@@ -27,6 +27,17 @@ You can create a test wallet containing multiple assets (including the base asse
 
+## Setting up a test wallet with multiple custom assets
+
+You can also create assets with specific `AssetId`s, coin amounts and number of coins.
```
```suggestion
You can also create assets with specific `AssetId`s, coin amounts, and number of coins.
```
I'm deeply sorry, it had to be done 😂 

![the-oxford-comma_52c855ed979ed_w1500](https://user-images.githubusercontent.com/3484025/181085624-782cbe70-42e8-4120-8c7b-f572682705a6.jpeg)

hal3e replied:
Will never forget it again :D

digorithm commented on packages/fuels-test-helpers/src/lib.rs:
```diff
@@ -91,6 +91,17 @@ pub fn setup_multiple_assets_coins(
     (coins, asset_ids)
 }
 
+/// Create a vector of of UTXOs with the provided AssetIds, num_coins and amount_per_coin
```
```suggestion
/// Create a vector of UTXOs with the provided AssetIds, num_coins, and amount_per_coin
```

digorithm commented on packages/fuels-test-helpers/src/wallets_config.rs:
```diff
@@ -1,34 +1,40 @@
 pub const DEFAULT_COIN_AMOUNT: u64 = 1_000_000_000;
 
+#[derive(Debug, Clone)]
+pub struct AssetsConfig {
```
How about calling it `AssetConfig` instead of `AssetsConfig` (notice the `s` gone)? This is the configuration for a single asset.

digorithm commented:
Adding the breaking label and editing the title to `feat!` so we can remember this is a breaking change (we're changing `WalletsConfig`'s pub methods) once we release it.

hal3e commented:
@digorithm thanks for the review!

digorithm reviewed (approved):
LGTM -- :shipit:

iqdecay commented on examples/wallets/src/lib.rs:
```diff
@@ -191,33 +191,37 @@ mod tests {
-        let asset_id_base = BASE_ASSET_ID;
-        let coins_per_asset_base = 2;
-        let amount_per_coin_base = 4;
+        let asset_base = AssetConfig {
```
`AssetConfig` is indeed much cleaner, like it!

hal3e replied:
Thanks!

iqdecay reviewed (approved):
Well done! :ship:
